aten/src/ATen/cuda/Atomic.cuh

```diff
@@ -402,6 +430,28 @@ __host__ __device__ T safe_max(T a, T b) {
   return max;
 }
 
+__inline__ __device__ c10::complex<float> complex_max(c10::complex<float> a, c10::complex<float> b) {
+  if(at::_isnan(b)) {
+    return b;
+  } else {
+    // Compute the magnitude of the complex numbers and compare each to see which one is greater.
+    float a_magnitude = __fsqrt_rn(
+      (
+        __fmul_rn(a.real(), a.real()) +
+        __fmul_rn(a.imag(),a.imag())
+      )
+    );
+    float b_magnitude = __fsqrt_rn(
+      (
+        __fmul_rn(b.real(), b.real()) +
+        __fmul_rn(b.imag(),b.imag())
+      )
+    );
+    return (a_magnitude > b_magnitude) ? a : b;
```

Comment:
Is there any precedence for this definition of complex max/min in PyTorch?

Reply:
I am not experienced enough with Pytorch to answer that. Aside from using magnitudes how else would you order them?

Reply:
Exactly, they cannot be ordered (in mathematical terms, complex numbers are not an [ordered field](https://en.wikipedia.org/wiki/Ordered_field)) 
We should error in these cases, same as we error when we call `max` on a complex tensor. If people want to use these ops on complex tensors, they can do a `view_as_real` and perform some transformations on the output to define the order they want.

Reply:
I was under the impression that some contexts will use magnitude for ordering complex numbers, like spectral analysis for DSP. I also took motivation from https://www.mathworks.com/help/matlab/ref/max.html as well

@Franklalalala 

Could you comment on whether or not you had a use case for scattering complex numbers? What kind of work were you trying to do? Would you know if ordering of complex numbers is practically useful?

Reply:
It may be implemented and it may be useful, but we don't implement that in PyTorch at a kernel level. As mentioned above, all these orderings can often be simulated with the current API and a bit of imagination :)

Reply:
I see - in that case I'll wait until @mikaylagawarecki has a chance to review again. Thanks for taking a look!

Reply:
@ZelboK I am working on [Tensor_network](https://en.wikipedia.org/wiki/Tensor_network), which require a series of matrix multiplication. In the case of complex elements, the torch scatter connot be used in GPU. As far as I concerned right now, we do not use sortage here, just elements' multiplication. 
By the way, we have worked a way out, that is, we transform the complex number through Euler transformation and turns the multiplication to addition of angles and multiplication of magnitude. 
The excellent work of you guys has reached out of my knowledge base, I connot give anymore advices. But thanks!

Reply:
> @ZelboK I am working on [Tensor_network](https://en.wikipedia.org/wiki/Tensor_network), which require a series of matrix multiplication. In the case of complex elements, the torch scatter connot be used in GPU. As far as I concerned right now, we do not use sortage here, just elements' multiplication. By the way, we have worked a way out, that is, we transform the complex number through Euler transformation and turns the multiplication to addition of angles and multiplication of magnitude. The excellent work of you guys has reached out of my knowledge base, I connot give anymore advices. But thanks!

Thanks a lot for responding, I was genuinely curious. This helps give me perspective :)